### ISSUE TITLE
[Enhancement]: High performance queries with hundreds of thousands of orders

### ISSUE THREAD

**Issue by ESPONI** [woothemes/woocommerce#35938]
### Describe the solution you'd like

Hi!,

the idea is to be able to delegate the counting of orders and line-items to SQL engine.

Our system manages more than 150.000 orders a year (read below) and something like 250.000 - 300.000 line-items (~2 per order as mean)

When using wc_get_orders or WC_Order_Query, though we use 'ids' parameter, the process is not efficient, as the counting is done in php:
$qt = count(wc_get_orders(...some_filtering_arguments...));

We propose including a parameter, as simple as 'count'
And the helper should prepare the SQL to delegate such task to SQL engine, returning only the result. 
(Improvements 18000 to 3 millis in our tests)

Also, it will be awesome to have:
wc_get_order_items or WC_Order_item_Query

Best regards!

### Describe alternatives you've considered

Prepare the queries internally as 'SELECT count()...'

### Additional context

**Trying to avoid TL;DR :)**

We sell physical products and tickets.
A ticket is “simply” a product that has a meta in order_item: is_ticket=1
Everything is ok until the real scenario:

150.000 ~ 160.000 orders/year.
Many orders has more than 1 line item, may be 5…
Some are products, 95% are tickets
Some tickets has another meta: ‘is_invitation=1’
Question:
How to perform thouse high performance queries using woocommerce classes and helpers, in a milisecond, like we do in SQL directly.
(We don’t want to lose compatibility, hardcoding SQL)

Basically, count total quantity sold of…

a given product_id
total invitations
and similar to those…

**Comment by ESPONI** [comment]
Hi!
i'm attaching a sample of our counting queries, let's see if there is any movement between WC Developers...

```
// Count SOLD Tickets (product_id aka ticket_id, for us)
function count_sold_tickets($ticket_id) { 
    global $wpdb;    
    
    $query = $wpdb->prepare('
        SELECT count(wc_order_item.order_item_id) as ticket_count
        FROM ' . $wpdb->prefix . 'woocommerce_order_items as wc_order_item
        INNER JOIN ' . $wpdb->prefix . 'woocommerce_order_itemmeta AS wc_order_item_meta ON wc_order_item.order_item_id = wc_order_item_meta.order_item_id
        INNER JOIN ' . $wpdb->prefix . 'posts AS wc_order ON wc_order_item.order_id = wc_order.ID
        
        WHERE 1=1
        AND wc_order_item_meta.meta_value = %d AND wc_order_item_meta.meta_key = "_product_id"
        AND (wc_order.post_status = "wc-completed" OR wc_order.post_status = "wc-on-hold");',
        (int) $ticket_id
    ); 

    $count_result = $wpdb->get_results($query);    

    if(count($count_result)) {
        return $count_result[0]->ticket_count;       
    } else {
        return 0;
    }
}```

**Comment by barryhughes** [comment]
Seems like a nice enhancement :+1:

**Comment by ESPONI** [comment]
> Seems like a nice enhancement 👍

Yes, we are using a more elaborated function, in our production project, based on that one, with several filters and variants.
In addition with this Index (see below), we do not understand very well why is woocommerce using the new tables for summarizing data.
It's all about good queries and good sql engine configuration (in order to efficiently cache data).
Honestly, where are woocommerce going?

ALTER TABLE `wp_stage`.`fMXeWxIt5_woocommerce_order_itemmeta` 
DROP INDEX `taicu_meta_key_value`, 
ADD INDEX `taicu_meta_key_value` (`meta_key`, `meta_value`(255)) 
USING BTREE;

**Comment by pmcpinto** [comment]
Thanks for such a detailed report, it's really helpful.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.
